packages/wallet/src/services/DelegationTracker/DelegationTracker.ts

```diff
@@ -0,0 +1,84 @@
+import { Cardano, StakePoolSearchProvider, WalletProvider } from '@cardano-sdk/core';
+import { CertificateType, Epoch } from '@cardano-sdk/core/src/Cardano';
+import { Delegation, Transactions } from '../types';
+import { KeyManager } from '../../KeyManagement';
+import { Observable, filter, map, share, switchMap } from 'rxjs';
+import { ObservableStakePoolSearchProvider, createDelegateeTracker, queryStakePoolsProvider } from './Delegatee';
+import { RetryBackoffConfig } from 'backoff-rxjs';
+import { RewardsHistoryProvider, createRewardsHistoryTracker } from './RewardsHistory';
+import { TrackerSubject, coldObservableProvider } from '../util';
+import { TxWithEpoch, transactionHasAnyCertificate } from './util';
+import { createRewardsHistoryProvider } from '.';
+
+export const createBlockEpochProvider =
+  (walletProvider: WalletProvider, retryBackoffConfig: RetryBackoffConfig) => (blockHashes: Cardano.Hash16[]) =>
+    coldObservableProvider(() => walletProvider.queryBlocksByHashes(blockHashes), retryBackoffConfig).pipe(
+      map((blocks) => blocks.map(({ epoch }) => epoch))
+    );
+
+export type BlockEpochProvider = ReturnType<typeof createBlockEpochProvider>;
+
+export interface DelegationTrackerProps {
+  walletProvider: WalletProvider;
+  keyManager: KeyManager;
+  stakePoolSearchProvider: StakePoolSearchProvider;
+  epoch$: Observable<Epoch>;
+  transactionsTracker: Transactions;
+  retryBackoffConfig: RetryBackoffConfig;
+  internals?: {
+    queryStakePools?: ObservableStakePoolSearchProvider;
+    rewardsHistoryProvider?: RewardsHistoryProvider;
+    blockEpochProvider?: BlockEpochProvider;
+  };
+}
+
+export const certificateTransactionsWithEpochs = (
+  transactionsTracker: Transactions,
+  blockEpochProvider: BlockEpochProvider,
+  certificateTypes: Cardano.CertificateType[]
+): Observable<TxWithEpoch[]> =>
+  transactionsTracker.history.outgoing$.pipe(
+    map((transactions) => transactions.filter((tx) => transactionHasAnyCertificate(tx, certificateTypes))),
+    // TODO
+    // distinctUntilChanged(transactionsEquals)
+    filter((transactions) => transactions.length > 0),
+    switchMap((transactions) =>
+      blockEpochProvider(transactions.map((tx) => tx.blockHeader.blockHash)).pipe(
+        map((epochs) => transactions.map((tx, txIndex) => ({ epoch: epochs[txIndex], tx })))
+      )
+    ),
+    share()
+  );
+
+// Review: all individual pieces of this are tested and it doesn't contain any logic
+// besides initializing the dependencies. Is it worth writing the test?
```

Comment:
I agree the test would just create more overhead with little benefit. I'm happy if it builds :wink: